hircluster.c

```diff
@@ -3781,6 +3701,132 @@ static void redisClusterAsyncCallback(redisAsyncContext *ac, void *r,
     cluster_async_data_free(cad);
 }
 
+/* Reply callback function for CLUSTER SLOTS */
+void clusterSlotsReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
+    UNUSED(ac);
+    redisClusterAsyncContext *acc = (redisClusterAsyncContext *)privdata;
+    redisClusterContext *cc = acc->cc;
+    acc->update_route_time = hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+
+    if (r == NULL) {
+        /* Ignore failures, next random node is hopefully working */
+        return;
+    }
+    redisReply *reply = (redisReply *)r;
+    dict *nodes = parse_cluster_slots(cc, reply, cc->flags);
+    if (updateNodesAndSlotmap(cc, nodes) != REDIS_OK) {
+        /* Ignore failures for now */
+    }
+}
+
+/* Reply callback function for CLUSTER NODES */
+void clusterNodesReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
+    UNUSED(ac);
+    redisClusterAsyncContext *acc = (redisClusterAsyncContext *)privdata;
+    redisClusterContext *cc = acc->cc;
+    acc->update_route_time = hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+
+    if (r == NULL) {
+        /* Ignore failures, next random node is hopefully working */
+        return;
+    }
+    redisReply *reply = (redisReply *)r;
+    dict *nodes = parse_cluster_nodes(cc, reply->str, reply->len, cc->flags);
+    if (updateNodesAndSlotmap(cc, nodes) != REDIS_OK) {
+        /* Ignore failures for now */
+    }
+}
+
+/* Return a random active node with async connection. */
+static redisClusterNode *getRandomActiveNode(dict *nodes) {
+    redisClusterNode *node;
+    dictIterator di;
+    dictInitIterator(&di, nodes);
+
+    /* Count nodes which we have a connection to */
+    int activeNodes = 0;
+    dictEntry *de;
+    while ((de = dictNext(&di)) != NULL) {
+        node = dictGetEntryVal(de);
+        if (node->acon != NULL && node->acon->err == 0 &&
+            node->acon->c.flags & REDIS_CONNECTED) {
+            activeNodes += 1;
+        }
+    }
+    /* Pick a random active node to return */
+    if (activeNodes > 0) {
+        int pick = (rand() % activeNodes) + 1; /* 1 to activeNodes */
+
+        int current = 0;
+        dictInitIterator(&di, nodes);
+        while ((de = dictNext(&di)) != NULL) {
+            node = dictGetEntryVal(de);
+            if (node->acon != NULL && node->acon->err == 0 &&
+                node->acon->c.flags & REDIS_CONNECTED) {
+                if (++current == pick) {
+                    return node;
+                }
+            }
+        }
+    }
+    return NULL;
+}
+
+/* Return a random node. */
+static redisClusterNode *getRandomNode(dict *nodes) {
+    dictIterator di;
+    dictInitIterator(&di, nodes);
```

Comment:
dict.c had support for picking random element, but maybe it was deleted from our copy of dict. The first few lines in dict.c mentions this:

```C
/* Hash table implementation.
 *
 * This file implements in memory hash tables with insert/del/replace/find/
 * get-random-element operations. Hash tables will auto resize if needed
```